src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -138,17 +139,51 @@ ss::future<walk_result> recursive_directory_walker::walk(
 
     fragmented_vector<ss::sstring> empty_dirs;
 
+    // Listing directories involves blocking I/O which in Seastar is serviced by
+    // a dedicated thread pool and workqueue. When listing directories with
+    // large number of sub-directories this leads to a large queue of work items
+    // for each of which we have to incur the cost of task switching. Even for a
+    // lightly loaded reactor, it can take up to 1ms for a full round-trip. The
+    // round-trip time is ~100x the syscall duration. With 2 level directory
+    // structure, 100K files, 2 getdents64 calls per directory this adds up to
+    // 400K syscalls and 6m of wall time.
+    //
+    // By running the directory listing in parallel we also parallelize the
+    // round-trip overhead. This leads to a significant speedup in the
+    // directory listing phase.
+    //
+    // Limit the number of concurrent directory reads to avoid running out of
+    // file descriptors.
+    //
+    // Empirical testing shows that this value is a good balance between
+    // performance and resource usage.
+    const size_t max_concurrency = 1000;
+
+    std::vector<ss::sstring> targets;
+    targets.reserve(max_concurrency);
+
     while (!state.empty()) {
-        auto target = state.pop();
-        vassert(
-          std::string_view(target).starts_with(start_dir),
-          "Looking at directory {}, which is outside of initial dir "
-          "{}.",
-          target,
-          start_dir);
-
-        co_await walker_process_directory(
-          start_dir, std::move(target), state, empty_dirs);
+        targets.clear();
+
+        auto concurrency = std::min(
+          state.dirlist.size(), size_t(max_concurrency));
+
+        for (size_t i = 0; i < concurrency; ++i) {
+            auto target = state.pop();
```

Comment:
nit: should we use `auto target = std::move(state.pop());`? Could probably apply `std::move` to [L77](https://github.com/redpanda-data/redpanda/blob/ee34998c5f67df6d6fdca1a9bcdcabd0b95ee6e9/src/v/cloud_storage/recursive_directory_walker.cc#L77) in this file as well.

Reply:
Adding std::move results in `Moving a temporary object prevents copy elision (fix available)clang(-Wpessimizing-move)`.

I like adding std::move to L77

Reply:
Huh, I'm surprised that's a pessimizing move!

Glad the compiler is wiser than I am.